---
title: "lunglps_hdWGCNA"
author: "Christian Heyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# single-cell analysis package
library(Seurat)

# plotting and data science packages
library(tidyverse)
library(cowplot)
library(patchwork)

# co-expression network analysis packages:
library(WGCNA)
library(hdWGCNA)
# using the cowplot theme for ggplot
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# optionally enable multithreading
#enableWGCNAThreads(nThreads = 4)
disableWGCNAThreads() 
```

## Test hdWGCNA

```{r}
#install.packages(c("Seurat", "WGCNA", "igraph", "devtools", "GeneOverlap"))

```

```{r}


```

```{r}
h5ad_path <- "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/merged/seurat/seurat_obj.rds.gz"
#h5_seuratpath <- 
  
  
#seurat_path <- Convert(source = h5ad_path, dest = "")

adata <- readRDS(h5ad_path)

adata@assays$counts <- NULL
adata@assays$soupX_counts <- NULL 
adata@assays$log1p_norm <- NULL
```

```{r}
DefaultAssay(adata) <- "RNA"
#adata@assays$scran_normalization@data[is.na(adata@assays$scran_normalization@data)] <- 0
#adata@assays$scran_normalization@counts[is.na(adata@assays$scran_normalization@counts)] <- 0
adata <- FindVariableFeatures(adata,assay = "RNA", selection.method = "vst", nfeatures = 4200)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(adata), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(adata,selection.method = "vst")
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 / plot2
```

```{r}
all.genes <- rownames(adata)
adata <- ScaleData(adata, features = all.genes)
adata <- RunPCA(adata) #a, features = VariableFeatures(object = adata))

print(adata[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(adata, dims = 1:2, reduction = "pca")
```

```{r}
SCpubr::do_DimPlot(adata, reduction = "pca",group.by = "sample") 
SCpubr::do_DimPlot(adata, reduction = "pca",group.by = "cell_type") 

DimHeatmap(adata, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(adata, dims = 1:15, cells = 500, balanced = TRUE)

ElbowPlot(adata)
```

```{r}

adata <- FindNeighbors(adata, dims = 1:15)
adata <- FindClusters(adata, resolution = 0.5)
#adata <- RunHarmony(adata, group.by.vars = "sample")
adata <- RunUMAP(adata, dims = 1:15,reduction = "pca")
```

```{r}
adata$cell_type[is.na(adata$cell_type)] <- "other"
SCpubr::do_DimPlot(adata,group.by = "sample")
SCpubr::do_DimPlot(adata,group.by = "cell_type")

SCpubr::do_DimPlot(adata, group.by = "seurat_clusters")
```


```{r}
adata <- SetupForWGCNA(
  adata,
  gene_select = "variable", # the gene selection approach
  fraction = 0.05, # fraction of cells that a gene needs to be expressed in order to be included
  wgcna_name = "lunglps" # the name of the hdWGCNA experiment
)

# construct metacells  in each group
adata <- MetacellsByGroups(
  seurat_obj = adata,
  group.by = c("cell_type", "sample"), # specify the columns in seurat_obj@meta.data to group by
  reduction = 'pca', # select the dimensionality reduction to perform KNN on
  k = 25, # nearest-neighbors parameter
  max_shared = 10, # maximum number of shared cells between two metacells
  ident.group = 'sample'# set the Idents of the metacell seurat object
)

# normalize metacell expression matrix:
adata <- NormalizeMetacells(adata)

```

## Check MEta cells

We check the low dimensional representation of the metacells to check how well they represent the dataset

```{r}
seurat_obj <- adata
metacell_obj <- GetMetacellObject(seurat_obj)

seurat_obj <- ScaleMetacells(seurat_obj, features=VariableFeatures(seurat_obj))
seurat_obj <- RunPCAMetacells(seurat_obj, features=VariableFeatures(seurat_obj))
seurat_obj <- RunHarmonyMetacells(seurat_obj, group.by.vars='sample')
seurat_obj <- RunUMAPMetacells(seurat_obj, reduction='harmony', dims=1:15)


p1 <- DimPlotMetacells(seurat_obj, group.by='cell_type') + umap_theme() + ggtitle("Cell Type")
p2 <- DimPlotMetacells(seurat_obj, group.by='sample') + umap_theme() + ggtitle("Sample")

rm(seurat_obj)

p1 | p2
```

```{r}
adata <- SetMultiExpr(
  adata,
  group_name = "immunec",
  group.by = "cell_type",
  multi.group.by ="sample",
  multi_groups = NULL, assay =  # this parameter can be used to select a subset of groups in the multi.group.by column,
  
  
)

consensus <- TRUE
```


```{r}
#adata <- SetDatExpr(
#  adata,
#  group_name = c("immunec", "devEC/proEC"), # the name of the group of interest in the group.by column
#  group.by='cell_type', # the metadata column containing the cell type info. This same column should have also been used in #MetacellsByGroups
#  assay = 'RNA', # using RNA assay
#  slot = 'data' # using normalized data,
  
#)

```

```{r}
# Test different soft powers:
if (consensus) {
  adata <- TestSoftPowersConsensus(adata)
} else {
  adata <- TestSoftPowers(
  adata,
  networkType = 'signed' # you can also use "unsigned" or "signed hybrid"
  )
}
# plot the results:
plot_list <- PlotSoftPowers(adata)


# get just the scale-free topology fit plot for each group
consensus_groups <- unique(adata$sample)
p_list <- lapply(1:length(consensus_groups), function(i){
  cur_group <- consensus_groups[[i]]
  plot_list[[i]][[1]] + ggtitle(paste0('Sex: ', cur_group)) + theme(plot.title=element_text(hjust=0.5))
})


# assemble with patchwork
wrap_plots(p_list, ncol=2)

power_table <- GetPowerTable(adata)
head(power_table)

pow <- c(5,5,4,6,6,6)
```



```{r}
# file.remove("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
library(bsub)
if (!file.exists("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")) {
  WGCNA_image <- "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData"
  save.image(WGCNA_image)
  bsub_chunk( name = "test hdWGCNA", packages = c("Seurat", "hdWGCNA", "WGCNA"),hours = 2, memory = 64, cores = 16, 
              R_version = "4.2.0",image = "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData", 
              working_dir = '/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps',
              code = {
  enableWGCNAThreads(nThreads = 16)
  adata <- ConstructNetwork(
  adata, soft_power=pow,
  setDatExpr=FALSE, tom_outdir = "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/data",
  consensus = TRUE,
  tom_name = 'immunec_consensus', saveConsensusTOMs = TRUE, overwrite_tom = TRUE # name of the topoligical overlap matrix written to disk
  )
  saveRDS(adata, file = "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
  })
}
adata <- readRDS("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
```


```{r}
library(hdWGCNA)
PlotDendrogram(adata, main='immunec')
```


Here Batch correction via harmony runs on the Module Eigengenes to deal with batch effects
```{r}
adata <- ScaleData(adata, features=VariableFeatures(adata))
#adata@assays$counts <- NULL
# compute all MEs in the full single-cell dataset

#adata <- RunUMAP(adata, dims = 1:15)

adata <- ModuleEigengenes(
 adata,
  
)

```


```{r}
# harmonized module eigengenes:
hMEs <- GetMEs(adata)

# module eigengenes:
MEs <- GetMEs(adata, harmonized=FALSE)
```

```{r}
adata <- ModuleConnectivity(
  adata,
  group.by = 'cell_type'
)

adata <- ModuleConnectivity(
  adata,
  group.by = 'sample', group_name = 'Day7'
)


adata <- ResetModuleNames(
  adata,
  new_name = "ectest-M"
)
p <- PlotKMEs(adata, ncol=5)
p
```

```{r}
modules <- GetModules(adata)

# show the first 6 columns:
modules %>%dplyr::filter(gene_name == "Aplnr")
```


```{r}
hub_df <- GetHubGenes(adata, n_hubs = 20)

hub_df
```


```{r}
library(UCell)
adata <- ModuleExprScore(
  adata,
  n_genes = 25,
  method='UCell'
)
```

```{r doplots}
# make a featureplot of hMEs for each module
plot_list <- ModuleFeaturePlot(
  adata,
  features='hMEs', # plot the hMEs
  order=TRUE # order so the points with highest hMEs are on top
)
library(patchwork)
# stitch together with patchwork
wrap_plots(plot_list, ncol=2)
```

```{r}
plot_list <- ModuleFeaturePlot(
  adata,
  features='scores', # plot the hub gene scores
  order='shuffle', # order so cells are shuffled
  ucell = TRUE # depending on Seurat vs UCell for gene scoring
)

wrap_plots(plot_list, ncol=2)

SCpubr::do_DimPlot(adata,group.by = "sample")
SCpubr::do_DimPlot(adata,group.by = "cell_type")
SCpubr::do_FeaturePlot(adata,features = "Aplnr",order = TRUE)

```


```{r}
# get hMEs from seurat object
MEs <- GetMEs(adata, harmonized=FALSE)
mods <- colnames(MEs); mods <- mods[mods != 'grey']

# add hMEs to Seurat meta-data:
adata@meta.data <- cbind(adata@meta.data, MEs)
SCpubr::do_FeaturePlot(adata,features = "ectest-M1",order = TRUE)
```

```{r}
# Plot INH-M4 hME using Seurat VlnPlot function
p <- VlnPlot(
  adata,
  features = 'ectest-M1',
  group.by = 'sample',
  pt.size = 0 # don't show actual data points
)

# add box-and-whisker plots on top:
p <- p + geom_boxplot(width=.25, fill='white')

# change axis labels and remove legend:
p <- p + xlab('') + ylab('hME') + NoLegend()

# plot output
p
```
```{r}
# Plot INH-M4 hME using Seurat VlnPlot function
p <- VlnPlot(
  adata,
  features = 'ectest-M2',
  group.by = 'sample',
  pt.size = 0 # don't show actual data points
)

# add box-and-whisker plots on top:
p <- p + geom_boxplot(width=.25, fill='white')

# change axis labels and remove legend:
p <- p + xlab('') + ylab('hME') + NoLegend()

# plot output
p
```
```{r}
ModuleCorrelogram(adata)
```

## Visualizing output

```{r}
# get hMEs from seurat object
MEs <- GetMEs(adata, harmonized=TRUE)
mods <- colnames(MEs); mods <- mods[mods != 'grey']

# add hMEs to Seurat meta-data:
adata@meta.data <- cbind(adata@meta.data, MEs)
```

```{r}
adata$cell_type[is.na(adata$cell_type)] <- "other"
# plot with Seurat's DotPlot function
p <- SCpubr::do_DotPlot(adata, features=mods, group.by = 'cell_type')

# flip the x/y axes, rotate the axis labels, and change color scheme:
p <- p +
  coord_flip() +
  RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue')

# plot output
p
```


### Module trait correlations

```{r}

# convert sex to factor
#adata$msex <- as.factor(adata$msex)

# convert age_death to numeric


# list of traits to correlate
cur_traits <- c('nCount_RNA', 'nFeature_RNA', 'total_counts_mt')

adata <- ModuleTraitCorrelation(
  adata,
  traits = cur_traits,
  group.by='cell_type'
)
```


```{r}
mt_cor <- GetModuleTraitCorrelation(adata)


```

```{r}
PlotModuleTraitCorrelation(
  adata,
  label = 'fdr',
  label_symbol = 'stars',
  text_size = 2,
  text_digits = 2,
  text_color = 'white',
  high_color = 'yellow',
  mid_color = 'black',
  low_color = 'purple',
  plot_max = 0.2,
  combine=TRUE
)
```



```{r}

seurat_obj <- adata
# change active hdWGCNA experiment to consensus
seurat_obj <- SetActiveWGCNA(seurat_obj, 'lunglps')

# compute eigengenes and connectivity
seurat_obj <- ModuleEigengenes(seurat_obj)
seurat_obj <- ModuleConnectivity(seurat_obj, group_name ='immunec', group.by='cell_type')

# visualize network with semi-supervised UMAP
seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 5,
  n_neighbors=5,
  min_dist=0.1,
  spread=2,
  wgcna_name = 'lunglps',
  target_weight=0.05,
  supervised=TRUE,
)

ModuleUMAPPlot(
  seurat_obj,
  edge.alpha=0.5,
  sample_edges=TRUE,
  keep_grey_edges=FALSE,
  edge_prop=0.075,
  label_hubs=0
)

```


