---
title: "lunglps_hdWGCNA"
author: "Christian Heyer"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 8, fig.height = 8)
consensus <- FALSE
# single-cell analysis package
library(Seurat)

# plotting and data science packages
library(tidyverse)
library(cowplot)
library(patchwork)
library(GeneOverlap)
# co-expression network analysis packages:++
library(WGCNA)
library(hdWGCNA)
# using the cowplot theme for ggplot
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# optionally enable multithreading
#enableWGCNAThreads(nThreads = 4)
disableWGCNAThreads() 
BPPARAM <- BiocParallel::SerialParam()
```

## Test hdWGCNA


```{r, eval=FALSE}
# get just the scale-free topology fit plot for each group
consensus_groups <- unique(adata$sample)
p_list <- lapply(1:length(consensus_groups), function(i){
  cur_group <- consensus_groups[[i]]
  plot_list[[i]][[1]] + ggtitle(paste0('Sex: ', cur_group)) + theme(plot.title=element_text(hjust=0.5))
})


# assemble with patchwork
wrap_plots(plot_list, ncol=2)

power_table <- GetPowerTable(adata)
head(power_table)

pow <- c(5,5,4,6,6,6)
```

```{r run_single, eval = FALSE}
# file.remove("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
library(bsub)
bsub_opt$call_Rscript = function(version) {
  GetoptLong::qq("module load gcc/11.1.0; module load curl/7.77.0; module load java/1.8.0_131; module load R/@{version}; Rscript")
}
if (!file.exists(glue::glue("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res_{net_type}.RDS.gz"))) {
  WGCNA_image <- "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData"
  save.image(WGCNA_image)
  bsub_chunk( name = "test_hdWGCNA", hours = 2, memory = 64, cores = 16,packages = c("hdWGCNA", "WGCNA", "Seurat"),
              R_version = "4.4.0",image = "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData", sh_head = "cd /omics/groups/OE0436/internal/heyer/projects/scRNAseq_scripts",
   #           working_dir = '/omics/groups/OE0436/internal/heyer/projects/scRNAseq_scripts',
              code = {
  #renv::activate(project = "/omics/groups/OE0436/internal/heyer/projects/scRNAseq_scripts")

  enableWGCNAThreads(nThreads = 16)
  adata <- ConstructNetwork(
  adata,  networkType = net_type, TOMType = net_type,
  setDatExpr=FALSE, tom_outdir = "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/data",
  consensus = FALSE,
  tom_name = 'immunec_signed', saveConsensusTOMs = TRUE, overwrite_tom = TRUE # name of the topoligical overlap matrix written to disk
  )
  saveRDS(adata, file = glue::glue("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res_{net_type}.RDS.gz"))
  })
}

```


```{r run_consensus, eval = FALSE}
# file.remove("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
library(bsub)
if (!file.exists("/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")) {
  WGCNA_image <- "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData"
  save.image(WGCNA_image)
  bsub_chunk( name = "test hdWGCNA", packages = c("Seurat", "hdWGCNA", "WGCNA"),hours = 2, memory = 64, cores = 16, 
              R_version = "4.4.0",image = "/omics/groups/OE0436/internal/heyer/temp/hdWGCNA_image.RData", 
              working_dir = '/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps', 
              code = {
  enableWGCNAThreads(nThreads = 16)
  adata <- ConstructNetwork(
  adata, soft_power=pow,
  setDatExpr=FALSE, tom_outdir = "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/data",
  consensus = TRUE,
  tom_name = 'immunec_consensus', saveConsensusTOMs = TRUE, overwrite_tom = TRUE # name of the topoligical overlap matrix written to disk
  )
  saveRDS(adata, file = "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res.RDS.gz")
  })
}

```


## Signed Netowrk analysis



```{r}
library(hdWGCNA)
signed_networks <-"/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res_signed.RDS.gz"
unsigned_networks <- "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/public_scrnaseq/lung_lps/hWCGNA_res_unsigned.RDS.gz"


downstream_analysis <- function(adata_path, main_title, group_by_vars, group_names, new_module_name) {
  adata <- readRDS(adata_path)
  
  
  adata <- ScaleData(adata, features=VariableFeatures(adata))
  adata <- ModuleEigengenes(adata, group.by.vars = group_by_vars)
  
  hMEs <- GetMEs(adata)
  MEs <- GetMEs(adata, harmonized=FALSE,wgcna_name = "lunglps")
  plot_list <- list()
  for ( group_n in group_names) {
    adata <- ModuleConnectivity(adata, group.by = 'cell_type', group_name = group_n)
    adata <- ResetModuleNames(adata, new_name = paste0(new_module_name, "_", group_n))
  
    p <- PlotKMEs(adata, ncol=3,)
    print(p)
    plot_list[[group_n]] <- p 
  }
  modules <- GetModules(adata)
  
  return(list(adata_obj = adata, hMEs=hMEs, MEs=MEs, modules=modules, kme_plots = plot_list ))
}
disableWGCNAThreads() 
signed_res <- downstream_analysis(signed_networks, "signed_network", group_by_vars = c("sample"), 
                                  group_names = c("EC"),
                                  new_module_name = "signed")


PlotDendrogram(signed_res$adata_obj, main="signed_network") 
signed_res$kme_plots$EC

```

### unsigned network

```{r}
unsigned_res <- downstream_analysis(unsigned_networks, "unsigned_network", 
                                   group_by_vars = c("sample"), 
                                  group_names = c("EC"),
                                  new_module_name = "unsigned")

PlotDendrogram(unsigned_res$adata_obj, main="unsigned_network") 

```

### Signed Network downstream analysis

```{r}
library(dplyr)
library(UCell)
library(patchwork)
library(SCpubr)
run_analysis <- function(res_list) {
  res_list$adata_obj -> adata
  res_list$modules -> modules
  # show the first 6 columns:
  
  modules %>% dplyr::filter(gene_name == "Aplnr")
  
  hub_df <- GetHubGenes(adata, n_hubs = 20)
  
  adata <- ModuleExprScore(
    adata,
    n_genes = 25,
    method='UCell'
  )
  
  # make a featureplot of hMEs for each module
  plot_list <- ModuleFeaturePlot(
    adata, reduction = "UMAP",
    features='hMEs', # plot the hMEs
    order=TRUE # order so the points with highest hMEs are on top
  )
  
  # stitch together with patchwork
  wrap_plots(plot_list, ncol=2) %>% print()
  
  plot_list <- ModuleFeaturePlot(
    adata,
    reduction = "UMAP",
    features='scores', # plot the hub gene scores
    order='shuffle', # order so cells are shuffled
    ucell = TRUE # depending on Seurat vs UCell for gene scoring
  )
  
  wrap_plots(plot_list, ncol=2) %>% print()
  
  SCpubr::do_DimPlot(adata,reduction = "UMAP",group.by = "sample")
  SCpubr::do_DimPlot(adata,reduction = "UMAP",group.by = "cell_type")
  SCpubr::do_FeaturePlot(adata,reduction = "UMAP",features = "Aplnr",order = TRUE)
  
  # get hMEs from seurat object
  MEs <- GetMEs(adata, harmonized=FALSE)
  mods <- colnames(MEs); mods <- mods[mods != 'grey']
  
  # add hMEs to Seurat meta-data:
  adata@meta.data <- cbind(adata@meta.data, MEs)
  #SCpubr::do_FeaturePlot(adata,reduction = "UMAP",features = "signed_EC1",order = TRUE)
  
  return(adata)
}

signed_res$adata_obj <- run_analysis(signed_res)

SCpubr::do_FeaturePlot(signed_res$adata_obj,reduction = "UMAP",features = "signed_EC1",order = TRUE)
SCpubr::do_FeaturePlot(signed_res$adata_obj,reduction = "UMAP",features = "Aplnr",order = TRUE)

p <- Seurat::FeatureScatter(signed_res$adata_obj, "Aplnr", "signed_EC1",group.by = "sample",plot.cor = T)
p
```

### Unsigned network downstream analysis 

```{r}
unsigned_res$adata_obj <- run_analysis(unsigned_res)
```

### Signed overview


```{r}
adata <- signed_res$adata_obj
# Plot INH-M4 hME using Seurat VlnPlot function

ME_names <- glue::glue("signed_EC{1:(ncol(adata@misc$lunglps$MEs) -1)}")
p <- SCpubr::do_ViolinPlot(
  adata,
  features = ME_names,
  group.by = 'sample',
  pt.size = 0, legend.position = "none" # don't show actual data points
)


# plot output
p
```



```{r}
ModuleCorrelogram(adata)
```

## Visualizing output

```{r}
# get hMEs from seurat object
MEs <- GetMEs(adata, harmonized=TRUE)
mods <- colnames(MEs); mods <- mods[mods != 'grey']

# add hMEs to Seurat meta-data:
adata@meta.data <- cbind(adata@meta.data, MEs)
```

```{r}
adata$cell_type[is.na(adata$cell_type)] <- "other"
# plot with Seurat's DotPlot function
p <- SCpubr::do_DotPlot(adata, features=mods, group.by = 'cell_type')

# flip the x/y axes, rotate the axis labels, and change color scheme:
p <- p +
  coord_flip() +
  RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue')

# plot output
p
```


### Unsigned overview

```{r}
#adata <- unsigned_res$adata_obj
# Plot INH-M4 hME using Seurat VlnPlot function

ME_names <- glue::glue("unsigned_EC{1:(ncol(adata@misc$lunglps$MEs) -1)}")
p <- SCpubr::do_ViolinPlot(
   unsigned_res$adata_obj,
  features = ME_names,
  group.by = 'sample',
  pt.size = 0, legend.position = "none" # don't show actual data points
)


# plot output
p
### Module trait correlations
```



```{r}
HubGeneNetworkPlot(
  signed_res$adata_obj,
  n_hubs = 5, n_other=10,
  edge_prop = 0.75,
  mods = 'all'
)
```


```{r}

seurat_obj <- adata
# change active hdWGCNA experiment to consensus
seurat_obj <- SetActiveWGCNA(seurat_obj, 'lunglps')

# compute eigengenes and connectivity
#seurat_obj <- ModuleEigengenes(seurat_obj)
#seurat_obj <- ModuleConnectivity(seurat_obj, group_name ='EC', group.by='cell_type')

# visualize network with semi-supervised UMAP
seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 10, # number of hub genes to include for the UMAP embedding
  n_neighbors=15, # neighbors parameter for UMAP
  min_dist=0.1 # min distance between points in UMAP space
)

ModuleUMAPPlot(
  seurat_obj,
  edge.alpha=0.25,
  sample_edges=TRUE,
  edge_prop=0.1, # proportion of edges to sample (20% here)
  label_hubs=2 ,# how many hub genes to plot per module?
  keep_grey_edges=FALSE

)

```

```{r enrichments, fig.width = 10}
library(enrichR)
# enrichr databases to test
dbs <- c('GO_Biological_Process_2023','GO_Cellular_Component_2023','GO_Molecular_Function_2023')

# perform enrichment tests
seurat_obj <- RunEnrichr(
  seurat_obj,
  dbs=dbs, # character vector of enrichr databases to test
  max_genes = Inf, # number of genes per module to test. use max_genes = Inf to choose all genes!,
  wgcna_name = "lunglps"
)

# retrieve the output table
enrich_df <- GetEnrichrTable(seurat_obj)
EnrichrBarPlot(
   seurat_obj,
  mods = "signed_EC1", # use all modules (this is the default behavior)
  database = "GO_Biological_Process_2023", # this has to be one of the lists we used above!!!
 n_terms= 10 # number of terms for each module
)
p <- EnrichrDotPlot(
   seurat_obj,
  mods = "all", # use all modules (this is the default behavior)
  database = "GO_Biological_Process_2023", # this has to be one of the lists we used above!!!
 n_terms= 3 # number of terms for each module
)
p
```

```{r}
# compute cell-type marker genes with Seurat:
Idents(seurat_obj) <- paste0(seurat_obj$sample, "_", seurat_obj$cell_type)
markers <- Seurat::FindAllMarkers(
  seurat_obj,
  only.pos = TRUE,
  logfc.threshold=1
)

# compute marker gene overlaps
overlap_df <- OverlapModulesDEGs(
  seurat_obj,
  deg_df = markers,
  fc_cutoff = 1 # log fold change cutoff for overlap analysis
)

plot_list <- OverlapBarPlot(overlap_df)

# stitch plots with patchwork
wrap_plots(plot_list[names(plot_list) %>% str_ends("EC")], ncol=3)
wrap_plots(plot_list[names(plot_list) %>% str_ends("Aerocyte")], ncol=3)
```

```{r}
OverlapDotPlot(
  overlap_df,
  plot_var = 'odds_ratio') +
  ggtitle('Overlap of modules & cell-type markers')
```

```{r}
dbs <- c("MSigDB_Hallmark_2020")
# perform enrichment tests
seurat_obj <- RunEnrichr(
  seurat_obj,
  dbs=dbs, # character vector of enrichr databases to test
  max_genes = Inf, # number of genes per module to test. use max_genes = Inf to choose all genes!,
  wgcna_name = "lunglps"
)

# retrieve the output table
enrich_df <- GetEnrichrTable(seurat_obj)
EnrichrBarPlot(
   seurat_obj,
  mods = "signed_EC1", # use all modules (this is the default behavior)
  database = "", # this has to be one of the lists we used above!!!
 n_terms= 10 # number of terms for each module
)
p <- EnrichrDotPlot(
   seurat_obj,
  mods = "all", # use all modules (this is the default behavior)
  database = "MSigDB_Hallmark_2020", # this has to be one of the lists we used above!!!
 n_terms= 3 # number of terms for each module
)
p
```

